security/authentication.md

```diff
@@ -0,0 +1,721 @@
+private:
+  std::string secret_;
+  std::string algorithm_;
+};
+```
+
+## API Key Authentication Implementation
+
+### API Key Authentication Checker
+```cpp
+#include <userver/server/handlers/auth/auth_checker_base.hpp>
+#include <userver/concurrent/variable.hpp>
+#include <userver/crypto/algorithm.hpp>
+
+class ApiKeyAuthChecker : public server::handlers::auth::AuthCheckerBase {
+public:
+  ApiKeyAuthChecker(const components::ComponentConfig& config,
+                    const components::ComponentContext& context)
+    : AuthCheckerBase(config, context) {
+    
+    // Load API keys from configuration or database
+    LoadApiKeys(config);
+  }
+  
+  AuthCheckResult CheckAuth(const server::http::HttpRequest& request,
+                           server::request::RequestContext& context) const override {
+    // Try different API key locations
+    std::string api_key;
+    
+    // 1. Check X-API-Key header
+    api_key = request.GetHeader("X-API-Key");
+    
+    // 2. Check Authorization header with API key scheme
+    if (api_key.empty()) {
+      auto auth_header = request.GetHeader("Authorization");
+      if (auth_header.starts_with("ApiKey ")) {
+        api_key = auth_header.substr(7);
+      }
+    }
+    
+    // 3. Check query parameter
+    if (api_key.empty()) {
+      api_key = request.GetArg("api_key");
+    }
+    
+    if (api_key.empty()) {
+      return AuthCheckResult::kTokenNotFound;
+    }
+    
+    // Validate API key
+    auto api_keys = api_keys_.Lock();
+    auto it = api_keys->find(api_key);
+    
+    if (it == api_keys->end()) {
+      RecordFailedAuth(request.GetRemoteAddress(), "invalid_api_key");
+      return AuthCheckResult::kTokenInvalid;
+    }
+    
+    const auto& key_info = it->second;
+    
+    // Check if key is active
+    if (!key_info.is_active) {
+      return AuthCheckResult::kTokenInvalid;
+    }
+    
+    // Check expiration
+    auto now = std::chrono::system_clock::now();
+    if (key_info.expires_at && now > *key_info.expires_at) {
+      return AuthCheckResult::kTokenExpired;
+    }
+    
+    // Check rate limits
+    if (!CheckRateLimit(api_key, key_info.rate_limit)) {
+      return AuthCheckResult::kRateLimited;
+    }
+    
+    // Set context data
+    context.SetData("api_key_id", key_info.id);
+    context.SetData("client_name", key_info.client_name);
+    context.SetData("permissions", key_info.permissions);
+    
+    RecordSuccessfulAuth(api_key);
+    
+    return AuthCheckResult::kOk;
+  }
+
+private:
+  struct ApiKeyInfo {
+    std::string id;
+    std::string client_name;
+    bool is_active = true;
+    std::optional<std::chrono::system_clock::time_point> expires_at;
+    std::vector<std::string> permissions;
+    int rate_limit = 1000; // requests per hour
+  };
+  
+  void LoadApiKeys(const components::ComponentConfig& config) {
+    auto keys_config = config["api-keys"];
+    auto keys = api_keys_.Lock();
+    
+    for (const auto& [key, info] : keys_config.Items()) {
```

Comment:
it's bad practise to store secret api keys in static config
if it is long-term key, it should be stored in secdist instead